### PR TITLE
Support for multiple devdiv instances in maestro

### PR DESCRIPTION
**Pull request by mmitche**
Also switch over to using MSI and KeyVault to grab secrets

### CODE REVIEW

**Comment by eerhardt** [comment]
Looks like CI is failing.  You broke a test.

**Comment by eerhardt** [comment]
> <?xml version="1.0" encoding="utf-8"?>

Is this necessary?  This is a web app.

---
Refers to: Maestro/src/Microsoft.DotNet.Maestro/app.config:1 in 4e4c0b4. [](commit_id = 4e4c0b431021f9446b82fcbf894fe6a5b8e27b14, deletion_comment = False)

File: Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs [hunk 1]
```diff
@@ -16,18 +16,21 @@ namespace Microsoft.DotNet.Maestro.Services
 {
     public class VsoService
     {
-        private static HttpClient client = new HttpClient();
+        private HttpClient client;
```
Comment (eerhardt) [inline Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs]: This is generally a bad change.  HttpClient instances shouldn't be created and then thrown away.

See [https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client#create-and-initialize-httpclient](https://docs.microsoft.com/en-us/aspnet/web-api/overview/advanced/calling-a-web-api-from-a-net-client#create-and-initialize-httpclient)

File: Maestro/src/Microsoft.DotNet.Maestro/Config.cs [hunk 2]
```diff
@@ -29,14 +32,39 @@ public string WebhookSecretToken
...
+            get { return _vstsInstanceUserName.Value; }
         }
 
-        public string VSoPassword
+        public string GetPassword(string instance)
         {
-            get { return _vsoPassword.Value; }
+            if (_vstsCreds.ContainsKey(instance))
+            {
+                return _vstsCreds[instance];
```
Comment (dagood) [inline Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: Is there any chance of multiple threads calling this?

If not, then I'd be a tiny bit happier with `TryGetValue` as a matter of style, although I'm sure perf doesn't matter here.

File: Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs [hunk 3]
```diff
@@ -38,7 +38,8 @@ public static ISubscriptionHandler TryCreate(JObject action, Subscription subscr
                     VsoProject = action["vsoProject"].ToString(),
                     BuildDefinitionId = action["buildDefinitionId"].Value<int>(),
                     SourceBranch = GetSourceBranch(subscription),
-                    VsoParameters = VsoParameterGenerator.GetParameters(subscription.ActionArguments, parameterSubstitutions)
+                    VsoParameters = VsoParameterGenerator.GetParameters(subscription.ActionArguments, parameterSubstitutions),
+                    VsoService = new VsoService(action["vsoInstance"].ToString())
```
Comment (dagood) [inline Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs]: This should be in `Execute`. Here, it'll work if there's no delay set, but if there's a delay, the class is JSON serialized and sent off to be executed later by the WebJob, where it's deserialized from JSON and this property is lost:

https://github.com/dotnet/versions/blob/95511b94d68bdd058b0c0574e7f2e4d8f39d1068/Maestro/src/Microsoft.DotNet.Maestro.WebJob/Functions.cs#L33-L34

(This is also causing the CI issue.)

File: Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs [hunk 4]
```diff
@@ -16,18 +16,21 @@ namespace Microsoft.DotNet.Maestro.Services
...
-        static VsoService()
+        public VsoService(String instance)
         {
+            this.instance = instance;
+            client = new HttpClient();
             client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic",
                Convert.ToBase64String(
                    Encoding.ASCII.GetBytes(
-                       string.Format("{0}:{1}", Config.Instance.VSoUser, Config.Instance.VSoPassword))));
+                       string.Format("{0}:{1}", Config.Instance.VstsUsername, Config.Instance.GetPassword(instance)))));
```
Comment (dagood) [inline Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs]: While you're changing this line, could you change it to string interpolation?

File: Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs (hunk 3 above)
Comment (mmitche) [reply Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs]: @dagood I think the CI issue is actually caused by the fact that it returned an empty string for credentials.

File: Maestro/src/Microsoft.DotNet.Maestro/Config.cs (hunk 2 above)
Comment (mmitche) [reply Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: @dagood I think you could get multiple threads if multiple handlers were called at once (actually, this should probably be a thread safe dictionary)

Comment (dagood) [reply Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: I'm a big fan of `ConcurrentDictionary`'s API, fwiw.

**Comment by mmitche** [comment]
> Is this necessary? This is a library.
> 
> Refers to: Maestro/src/Microsoft.DotNet.Maestro/app.config:1 in 4e4c0b4. [](commit_id = 4e4c0b4, deletion_comment = False)
> 

@eerhardt Are you referring to the general inclusion of app.config?  Honestly I'm not sure.  VS added the binding redirects automatically because there were mismatched versions in one of the new nuget packages required for the keyvault changes.  The webapp wouldn't load locally otherwise.

File: Maestro/src/Microsoft.DotNet.Maestro/Config.cs [hunk 5]
```diff
@@ -15,10 +18,10 @@ public class Config
         public static Config Instance { get; } = new Config();
 
         private Lazy<string> _webhookSecretToken = new Lazy<string>(() => ConfigurationManager.AppSettings[nameof(WebhookSecretToken)]);
-        private Lazy<string> _vsoUser = new Lazy<string>(() => ConfigurationManager.AppSettings[nameof(VSoUser)]);
-        private Lazy<string> _vsoPassword = new Lazy<string>(() => ConfigurationManager.AppSettings[nameof(VSoPassword)]);
         private Lazy<string> _subscriptionsUrl = new Lazy<string>(() => ConfigurationManager.AppSettings[nameof(SubscriptionsUrl)]);
         private Lazy<string> _azureWebJobsStorage = new Lazy<string>(() => ConfigurationManager.ConnectionStrings[nameof(AzureWebJobsStorage)]?.ConnectionString);
+        private Lazy<string> _vstsInstanceUserName = new Lazy<string>(() => ConfigurationManager.AppSettings[nameof(VstsUsername)]);
+        private Dictionary<string, string> _vstsCreds = new Dictionary<string, string>();
```
Comment (jaredpar) [inline Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: Are instance names case sensitive?

File: Maestro/src/Microsoft.DotNet.Maestro/Config.cs [hunk 6]
```diff
@@ -29,14 +32,39 @@ public string WebhookSecretToken
...
-            get { return _vsoPassword.Value; }
+            if (_vstsCreds.ContainsKey(instance))
+            {
+                return _vstsCreds[instance];
+            }
+            else
+            {
+                // Grab the secret name from the web.config
+                string secretName = ConfigurationManager.AppSettings[$"{instance}-KeyVaultSecretName"];
+                if (String.IsNullOrEmpty(secretName))
```
Comment (jaredpar) [inline Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: string

File: Maestro/src/Microsoft.DotNet.Maestro/Microsoft.DotNet.Maestro.csproj [hunk 7]
```diff
@@ -11,6 +11,8 @@
     <AssemblyName>Microsoft.DotNet.Maestro</AssemblyName>
     <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
     <FileAlignment>512</FileAlignment>
+    <NuGetPackageImportStamp>
+    </NuGetPackageImportStamp>
```
Comment (jaredpar) [inline Maestro/src/Microsoft.DotNet.Maestro/Microsoft.DotNet.Maestro.csproj]: What does this entry do?

File: Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs (hunk 3 above)
Comment (dagood) [reply Maestro/src/Microsoft.DotNet.Maestro/Handlers/VsoBuildHandler.cs]: I mean that at a higher level, CI shouldn't be trying to get credentials and moving it to `Execute` would fix that.

File: Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs [hunk 8]
```diff
@@ -16,18 +16,21 @@ namespace Microsoft.DotNet.Maestro.Services
 {
     public class VsoService
     {
-        private static HttpClient client = new HttpClient();
+        private HttpClient client;
+        private string instance;
```
Comment (jaredpar) [inline Maestro/src/Microsoft.DotNet.Maestro/Services/VsoService.cs]: Rest of change uses `_` prefix on private instance fields

File: Maestro/src/Microsoft.DotNet.Maestro/Config.cs (hunk 5 above)
Comment (mmitche) [reply Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: In our subscriptions.json they are all lower case.  Maybe worth changing to a case insensitive comparer though.  Are Configuration Manager keys insensitive?

Comment (jaredpar) [reply Maestro/src/Microsoft.DotNet.Maestro/Config.cs]: I don't know off hand but my guess is it's case insensitive. 

---
In reply to: [190612202](https://github.com/dotnet/versions/pull/296#discussion_r190612202) [](ancestors = 190612202)

**Comment by eerhardt** [comment]
> @eerhardt Are you referring to the general inclusion of app.config?

Yes.  I don't think the app.config for the library project is necessary and should be removed.

> The webapp wouldn't load locally otherwise.

The changes in the Web.config fix that.  But the library's app.config is what should be reverted IMO.

**Comment by mmitche** [comment]
Fixed:
- Reverted HttpClient back to status (moved auth into POST)
- ConcurrentDictionary

**Comment by mmitche** [comment]
@eerhardt Any more comments?

**Comment by dagood** [comment]
Does https://github.com/dotnet/versions/pull/303 depend on this? (Is it already deployed?)

**Comment by mmitche** [comment]
@dagood Deployed as of this morning.

**Pull request merged**